src/FHE.sol

```diff
@@ -856,6 +997,42 @@ contract FHE {
         );
     }
 
+    /// Encrypt a plaintext uint64 value
+    /// @dev This costs 1000 gas
+    /// @param plaintextValue Plaintext value to multiply to a ciphertext
+    /// @return result The encrypted `plaintextValue`
+    function encryptUint64(uint64 plaintextValue) public view returns (bytes memory) {
+        bytes memory msg_sender = uint256(uint160(msg.sender)).toBytes(32);
+        bytes memory block_number = block.number.toBytes(32);
+
+        bytes memory publicData = bytes.concat(msg_sender, block_number);
+
+        bytes memory input = packUint64Bytes(plaintextValue, publicData);
+        return callPrecompile(ENCRYPT_UINT64_ADDRESS, ENCRYPT_GAS, input);
+    }
+
+    /// Reencrypt a ciphertext from the network key to another key.
+    /// @dev This costs 2000 gas
+    /// @param pubk Bincode encoded `Sunscreen::PublicKey`
+    /// @param encryptedValue Bincode encoded `Sunscreen::Cipher<bfv::Unsigned256>`.
+    /// @return result The reencrypted `encryptedValue` under the provided key.
+    function reencryptUint64(bytes memory pubk, bytes memory encryptedValue) public view returns (bytes memory) {
+        bytes memory msg_sender = uint256(uint160(msg.sender)).toBytes(32);
+        bytes memory block_number = block.number.toBytes(32);
+
+        bytes memory publicData = bytes.concat(msg_sender, block_number);
```

Comment:
For the reencrypt case, the private value is used in the hash as well. For regular encryption, a random (but constant) value is hashed in to have some private information.

There isn't a ton we can do without a consistent random number generator between sealers :/. But this is mostly just for demonstrating the API, not the full threshold implementation yet.

Reply:
> For the reencrypt case, the private value is used in the hash as well. For regular encryption, a random (but constant) value is hashed in to have some private information.

This doesn't address the point above. I'm talking about a single user rencrypting two ciphertexts that _happen_ to be the same value. E.g. the user put up two yes/no polls and is getting their results. If this happens in the same block, the fact that the poll outcomes are equal is revealed.

I agree it's going to maybe be whack-a-mole trying to cover every case just by hashing things. It might mean though that we should expose methods that allow the user to pass in public data for hashing, if they have some use case that isn't secure with our chosen hashing methods.

Reply:
Well I had thought about also adding the transaction origin but I don't think that addresses your case. It's nice that the reencrypt is a pure function but that also means unless we can get solidity to provide something unique per transaction we will get the same result :/